### ISSUE TITLE
reproducible results

### ISSUE THREAD

**Issue by davide1993** [autonomio/talos#175]
I don't know how to get reproducible results with talos. I run a scan whit some params and If run the test with the same params I get the same file results.csv and that's ok. Now if I want to get the best model from the results file.csv I pick up the params from csv file and run the scan agasin  (only 1 permutation in this case) but the results are different from before. Moreover,  when I try to restore a model,  Restore.results variable contains only the first two lines of the csv file and not all the others. Tell me please how to save models of a massive test and later rerun the best model I choose.
Thank you.

-------------------------------------------------------------------------
SCAN TEST:

```
scan = ta.Scan(X_train, y_train,
            params=params,
            dataset_name="../test/"+dataset.name,
            x_val=X_test,
            y_val=y_test,
            experiment_no=date,
            model=dataset.getCNNModel,
            grid_downsample=1)
```

PARAMS:
```
        params = {'lr': (2, 10, 5),
            'first_neuron':[8,16,32,64],
            'hidden_layers': [1,2, 3, 4,5,6],
            'batch_size': [16,32,64,128],
            'epochs': [100,200,300],
            'dropout': (0, 0.40, 10),
            'weight_regulizer': [None],
            'emb_output_dims': [None],
            'optimizer': ['Adam', 'Nadam', 'RMSprop'],
            'losses': [binary_crossentropy],
            'activation': [relu,elu],
            'last_activation': [sigmoid]}

```

 ```
def getCNNModel(self, x_train, y_train, x_val, y_val, params):
        
        model = Sequential()
        model.add(Dense(params['first_neuron'],
                        input_dim=x_train.shape[1],
                        activation=params['activation']))
        #model.add(Dropout(params['dropout']))
        model.add(Dense(64, activation='relu'))
        model.add(Dense(8, activation='relu'))
        model.add(Dense(1, activation=params['last_activation']))
    
        model.compile(optimizer=params['optimizer'],#(lr_normalizer(params['lr'], params['optimizer'])),
                      loss=params['losses'],
                      metrics=['acc',fmeasure_acc,precision_acc, recall_acc])
    
        history = model.fit(x_train, y_train,
                        batch_size=params['batch_size'],
                        epochs=params['epochs'],
                        verbose=0,
                        validation_data=[x_val, y_val])
                        #callbacks=[live()])  da aggiungere se si vuole vedere il plotting durante le epoche ma molto leto
    
        return history, model
  
```

**Comment by mikkokotila** [comment]
Firstly, whenever you train a model, you will get variation, because deep learning is not deterministic. 

Regarding your use case, Talos provides several possibilities in terms of getting the best model as a result of cross-validation. The easiest is to just use the `scan_object.best_model()` if all you want to do is get the best model of the Scan() results. You don't ever need to go to the csv, that's just for reporting and manual analysis purpose. 

If you want to move to another system, or just want to take things offline and come back later, you can do: 

`ta.Deploy(scan_object, 'your_model_name')`

**Comment by davide1993** [comment]
I've allready tried Deploy but when I restore the object I think something goes wrong: 

- deploy.results does not contains the whole results dataframe  of the previous scan but only the first and second row of dataframe (of course I check the file stored and is ok, the problem is  when restoring).

- hidden layer's array of restored object contains more Dense element that the best model has


I don't understand what do you mean when you say deep learning isn't deterministic. I set random seed, disable multiple threads so it should be deterministic (https://keras.io/getting-started/faq/).
When and where should I specify the metrics in order to save the best model of that metric, moreover what is the right way to store different models for different metrics? For example, after a scan I want to save the best model for recall, the best model for accuracy and so on... The right procedure is to call deplay one time for even metric? The saved objects when restore looks the same. To save different models I run the code below:
```
Deploy(scan_object, 'experiment_name', 'val_acc') #store the best accuracy model
Deploy(scan_object, 'experiment_name', 'val_fmeasure_acc') #store the best fmeasure 
....
```

**Comment by mikkokotila** [comment]
@davide1993 I can see from my email that you replied, but I can't see the reply here. Did you delete it?

**Comment by davide1993** [comment]
yes because I find now an error in my code.
If I have the same problem after fixed this I'll tell you, thanks.

**Comment by mikkokotila** [comment]
That's great :) Closing here. Feel feel to open a new issue in case something comes up.
